File: BookReaderObjC/BookReaderObjC/ViewController.m
//
//  ViewController.m
//  BookReaderObjC
//
//  Created by Field Employee on 3/27/20.
//  Copyright © 2020 Hugo Flores. All rights reserved.
//

#import "ViewController.h"
#import "PageContentViewController.h"

@interface ViewController () <UIPageViewControllerDataSource>
@property NSArray *arrPageTitle;
@property NSArray *arrPagePhoto;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    _arrPageTitle = @[@"This is The App", @"This is a View", @"This is a second view"];
    _arrPagePhoto = @[@"voight", @"apple", @"Cyberia"];
    
    self.dataSource = self;

    [self setViewControllers:@[[self getViewControllerAtIndex:0]] direction:UIPageViewControllerNavigationDirectionForward animated:false completion:nil];

}

- (PageContentViewController *) getViewControllerAtIndex:(NSInteger)index {
    PageContentViewController *pageContentViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"PageContentViewController"];

    pageContentViewController.strTitle = _arrPageTitle[index];
    pageContentViewController.strPhotoName = [NSString stringWithFormat:@"%@", _arrPagePhoto[index]];
    pageContentViewController.pageIndex = (int) index;
    printf("Title = %s\n", [pageContentViewController.strTitle UTF8String]);
    return pageContentViewController;
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    PageContentViewController *pageContent = (PageContentViewController *)viewController;
    int index = pageContent.pageIndex;
    
    if (index == NSNotFound) {
        return nil;
    }
    
    index += 1;
    if (index == _arrPageTitle.count){
        return nil;
    }
    
    return [self getViewControllerAtIndex:index];
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    PageContentViewController *pageContent = (PageContentViewController *)viewController;
    int index = pageContent.pageIndex;
    
    if ((index == 0) || (index == NSNotFound)) {
        return nil;
    }
    
    index -= 1;
    
    return [self getViewControllerAtIndex:index];
}

@end

File: PageVC/UIPageViewControllerDemo/Controller/ViewController.swift
//
//  ViewController.swift
//  UIPageViewControllerDemo
//
//

import UIKit

class ViewController: UIPageViewController, UIPageViewControllerDataSource
{
    var arrPageTitle: NSArray = NSArray()
    var arrPagePhoto: NSArray = NSArray()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        arrPageTitle = ["This is The App", "This is Second", "This is Third"];
        arrPagePhoto = ["apple.png", "Swift.png", "objc.png"];
        
        self.dataSource = self
        
        self.setViewControllers([getViewControllerAtIndex(index: 0)] as [UIViewController], direction: UIPageViewController.NavigationDirection.forward, animated: false, completion: nil)
    }

// MARK:- UIPageViewControllerDataSource Methods
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        let pageContent: PageContentViewController = viewController as! PageContentViewController
        
        var index = pageContent.pageIndex
        
        if ((index == 0) || (index == NSNotFound)) {
            return nil
        }
        
        index-=1;
        return getViewControllerAtIndex(index: index)
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        let pageContent: PageContentViewController = viewController as! PageContentViewController
        
        var index = pageContent.pageIndex
        
        if (index == NSNotFound){
            return nil;
        }
        
        index+=1;
        if (index == arrPageTitle.count){
            return nil;
        }
        return getViewControllerAtIndex(index: index)
    }

// MARK:- Other Methods
    func getViewControllerAtIndex(index: NSInteger) -> PageContentViewController
    {
        // Create a new view controller and pass suitable data.
        let pageContentViewController = self.storyboard?.instantiateViewController(withIdentifier: "PageContentViewController") as! PageContentViewController

        pageContentViewController.strTitle = "\(arrPageTitle[index])"
        pageContentViewController.strPhotoName = "\(arrPagePhoto[index])"
        pageContentViewController.pageIndex = index
        
        return pageContentViewController
    }
}

